_blogposts/2023-04-15-improving-interop.mdx

```diff
@@ -0,0 +1,317 @@
+
+The same logic is easily applied to string literals from TypeScript, only here the benefits is even larger, because string literals have the same limitations in TypeScript that polymorphic variants have in ReScript.
+
+```typescript
+// direction.ts
+type direction = "UP" | "DOWN" | "LEFT" | "RIGHT";
+```
+
+There's no way to attach documentation strings to string literals in TypeScript, and you only get the actual value to interact with.
+
+With the new customizable variants, you could bind to the above string literal type easily, but add documentation, and change the name you interact with in ReScript. And there's no runtime cost.
+
+### Untagged variants
+
+We've also implemented support for _untagged variants_. This will let you use variants to represent values that are primitives and literals in a way that hasn't been possible before.
+
+We'll explain what this is and why it's useful by showing a number of real world examples. Let's start with a simple one on how we can now represent a heterogenous array.
+
+```rescript
+@unboxed type listItemValue = String(string) | Boolean(bool) | Number(float)
+
+let myArray = [String("Hello"), Boolean(true), Boolean(false), Number(13.37)]
+```
+
+Here, each value will be _unboxed_ at runtime. That means that the variant payload will be all that's left, the variant case name wrapping the payload itself will be stripped out and the payload will be all that remains.
+
+It, therefore, compiles to this JS:
+
+```javascript
+var myArray = ["hello", true, false, 13.37];
+```
+
+This was previously possible to do, leveraging a few tricks, when you didn't need to potentially read the values from the array again in ReScript. But, if you wanted to read back the values, you'd have to do a number of manual steps.
+
+In the above example, reaching back into the values is as simple as pattern matching on them.
+
+Let's look at a few more examples of what untagged variants enable.
+
+### Pattern matching on nullable values
+
+Previously, any value that might be `null` would need to be explicitly converted to an option by using for example `Nullable.toOption` before you could use pattern matching on it. Here's a typical example of how that could look:
+
+```rescript
+type userAge = {ageNum: Nullable.t<int>}
+
+type rec user = {
+  name: string,
+  age: Nullable.t<userAge>,
+  bestFriend: Nullable.t<user>,
+}
+
+let getBestFriendsAge = user =>
+  switch user.bestFriend->Nullable.toOption {
+  | Some({age}) =>
+    switch age->Nullable.toOption {
+    | None => None
+    | Some({ageNum}) => ageNum->Nullable.toOption
+    }
+  | None => None
+  }
+```
+
+As you can see, you need to convert each level of nullables explicitly, which makes it hard to fully utilize pattern matching. With the new unboxed variant representation, we'll instead be able to do this:
+
+```rescript
+// The type definition below is inlined here to examplify, but this definition will live in Core and be easily accessible
+module Nullable = {
```

Comment:
The name `Nullable` may be confusing because `Js.Nullable.t` and Core's `Nullable.t` can also be `undefined`, not just `null`.

Reply:
Good point! Will change.